### ISSUE TITLE
Statement timeout produces cryptic error messages for end users (`unexpected`, `database query error`)

### ISSUE THREAD

**Issue by bitjson** [hasura/graphql-engine#7839]
### Is your proposal related to a problem?

For Hasura deployments which use [`statement_timeout` Database Timeouts](https://hasura.io/learn/graphql/hasura-advanced/security/6-timeouts/): if the timeout is reached, end users receive a confusing error message.

Postgres returns an error like `canceling statement due to statement timeout`, but Hasura returns a generic error:
```json
{
  "errors": [
    {
      "extensions": {
        "path": "$",
        "code": "unexpected"
      },
      "message": "database query error"
    }
  ]
}
```

Most new users of our API think this message means either:
- **their GraphQL syntax is incorrect** and they spend time trying to fix, becoming frustrated when the error is non-deterministic (because the DB manages to run the query faster once some data is in memory)
- **our Hasura server is unreliable or broken**

### Describe the solution you'd like

It would be fantastic if Hasura identified statement timeout errors and instead sent a more helpful error back to the user:
```JSON
{
  "errors": [
    {
      "message": "Request cancelled due to excessive resource usage. Please limit your query and try again."
    }
  ]
}
```

### If the feature is approved, would you be willing to submit a PR?

I'd be happy to try! I'm very new to Haskell though, so guidance would be appreciated.

**Comment by Bessonov** [comment]
If it's OK to expose more information, then you can try to activate [dev-mode](https://hasura.io/docs/latest/graphql/core/deployment/graphql-engine-flags/reference.html).This should give you a full message.

But it doesn't help me with my issue. I use serializable transaction isolation mode and must handle 40001 error (could not serialize access due to concurrent update). Without activating dev mode I don't see any way how to tell the client to handle the error.

Probably, it would be better to allow devs handle errors and output custom messages.

Related: https://github.com/hasura/graphql-engine/issues/5626

**Comment by Bessonov** [comment]
Another relevant error is `40P01 | deadlock_detected`.

**Comment by jflambert** [comment]
I strongly agree with improving this error handling. It should be as simple as catching SQL state `57014` and displaying an appropriate message.
